test/jdk/java/nio/jni/NewDirectByteBuffer.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+/*
+ * @test
+ * @bug 8299684
+ * @summary Verify that JNI NewDirectByteBuffer throws
+ * IllegalArgumentException if the capacity is negative or greater than
+ * Integer::MAX_VALUE
+ * @requires (sun.arch.data.model == "64" & os.maxMemory >= 8g)
+ * @run junit/othervm/native NewDirectByteBuffer
+ */
+public class NewDirectByteBuffer {
+    static {
+        System.loadLibrary("NewDirectByteBuffer");
+    }
+
+    private static final void checkBuffer(ByteBuffer buf, long capacity) {
+        Assertions.assertTrue(buf.isDirect(), "Buffer is not direct");
+        Assertions.assertEquals(capacity, getDirectByteBufferCapacity(buf),
+            "GetDirectBufferCapacity returned unexpected value");
+        Assertions.assertEquals(capacity, buf.capacity(),
+            "Buffer::capacity returned unexpected value");
+        Assertions.assertEquals(0L, buf.position(),
+            "Buffer::position returned unexpected value");
+        Assertions.assertEquals(capacity, buf.limit(),
+            "Buffer::limit returned unexpected value");
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {0L, 1L, (long)Integer.MAX_VALUE/2,
+        (long)Integer.MAX_VALUE - 1, (long)Integer.MAX_VALUE})
+    void legalCapacities(long capacity) {
+        try {
+            final AtomicReference<ByteBuffer> bufHolder = new AtomicReference();
+            Assertions.assertDoesNotThrow(() -> {
+                try {
+                    bufHolder.set(newDirectByteBuffer(capacity));
+                } finally {
+                    ByteBuffer buf = bufHolder.get();
+                    if (buf != null) {
+                        freeDirectByteBufferMemory(buf);
+                    }
+                }
+            });
+            ByteBuffer buf = bufHolder.get();
+            if (buf != null) {
+                checkBuffer(buf, capacity);
+            }
+        } catch (OutOfMemoryError ignored) {
+            // Ignore the error
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {Long.MIN_VALUE, (long)Integer.MIN_VALUE - 1L, -1L,
+        (long)Integer.MAX_VALUE + 1L, 3_000_000_000L, 5_000_000_000L,
+        Long.MAX_VALUE})
+    void illegalCapacities(long capacity) {
+        try {
+            Assertions.assertThrows(IllegalArgumentException.class, () -> {
+                ByteBuffer buf = newDirectByteBuffer(capacity);
+                if (buf != null) {
+                    freeDirectByteBufferMemory(buf);
+                }
+            });
+        } catch (OutOfMemoryError ignored) {
```

Comment:
Apparently it is not needed so removed in 38ea7997be2be59e2cea97c7dabf929210c0edfa.

Reply:
No wonder I couldn't find this change! The catch of OOME was added because I still believe that trying to allocate a buffer of capacity Integer.MAX_VALUE will throw OOME (at least some of the time). If that happens then the test should not fail.

Reply:
No objection to having the test pass if malloc fails but it would be better if the native newDirectByteBuffer were changed to just call JNI NewDirectByteBuffer with a given address so that the malloc + JNI function aren't in the same native method. That allows the test to distinguish malloc returning NULl from NewDirectByteBuffer throwing OOME. The native method freeDirectByteBufferMemory can be split too so that there is native method get the direct buffer address. With these changes, the test can use Unsafe allocateMemory/freeMemory (changing to FFM APIs in the future) or have native methods that do the malloc/free. The more basic methods means the test can be improved,  e.g. checkBuffer doesn't currently test that GetDirectBufferAddress returns the address that the DBB was created to wrap.